### PR TITLE
add CA Cert functionality

### PR DESCRIPTION
**Pull request by nivibilla**
WIP. Code looks fine to me but still getting errors.

### CODE REVIEW

**Comment by nivibilla** [comment]
hi,

Im trying to use a CA certificate when building a proxy. The python equivalent im currently using which works is:

```
import openai
import httpx

certificate_path = './CA.crt'

openai_client = openai.OpenAI(
    api_key='api_key',
    base_url='base_url',
    http_client=httpx.Client(verify=certificate_path)
)
```

Ive tried to use the ca_certs param in uvicorn but I keep getting this error

<details>

`ssl.SSLCertVerificationError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:997)`

`raise OpenAIError(status_code=500, message=f"{str(e)}")
litellm.llms.openai.OpenAIError: Connection error.`

`raise ValueError("No models available")`
</details>

**Comment by ishaan-jaff** [comment]
@nivibilla can you clarify
- are you trying to use the litellm proxy server?
- are you trying to use `litellm.completion` and set `certificate_path`

**Comment by nivibilla** [comment]
I'm trying to use the proxy server. With the certificate 

Pls see the attempted changes to the proxy cli file.

File: litellm/proxy/proxy_cli.py [hunk 1]
```diff
@@ -502,6 +510,16 @@ def _make_openai_completion():
...
+            elif ssl_ca_bundle_path is not None:
+                print(
+                    f"\033[1;32mLiteLLM Proxy: Using SSL with CA Bundle: {ssl_ca_bundle_path}\033[0m\n"
+                )
+                uvicorn.run(
+                    app,
+                    host=host,
+                    port=port,
+                    ssl_ca_certs=ssl_ca_bundle_path,
+                )  # run uvicorn
```
Comment (ishaan-jaff) [inline litellm/proxy/proxy_cli.py:522]: this looks good to me, just need to check uvicorn supports `ssl_ca_certs` and gunicorn supports `ca_certs` 

@nivibilla can you confirm the proxy works on your side with these changes ?

**Review by ishaan-jaff — CHANGES_REQUESTED** [review]
Good to merge after @nivibilla can confirm the changes work on their side

**Comment by ghost** [comment]
Based on conversation, I don't think this PR would solve the problem you faced @nivibilla 

Instead we can just read the ca bundle path from the environment and set it for the client

**Comment by ishaan-jaff** [comment]
@nivibilla can you try setting `verify=your ca bundle path` here and can you lmk if it works ? https://github.com/BerriAI/litellm/blob/1391490d928f32496298ee002d13d068e4e80b7d/litellm/router.py#L1672  - i'm assuming your models are OpenAI compatible 


```python
_client = openai.AsyncOpenAI(  # type: ignore
                    api_key=api_key,
                    base_url=api_base,
                    timeout=timeout,
                    max_retries=max_retries,
                    organization=organization,
                    http_client=httpx.AsyncClient(
                        transport=AsyncCustomHTTPTransport(),
                        limits=httpx.Limits(
                            max_connections=1000, max_keepalive_connections=100
                        ),
                        verify="your_ca_bundle_path",
                        mounts=async_proxy_mounts,
                    ),  # type: ignore
                )
```

**Comment by ishaan-jaff** [comment]
did the fix work @nivibilla ?

**Comment by nivibilla** [comment]
Unfortunately not. 

Same error 
`ssl.SSLCertVerificationError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:997)`

**Comment by SebastienLentzAM** [comment]
I've been experiencing the same issue (https://docs.litellm.ai/docs/proxy/deploy#2-setting-ssl-certification didn't work in my case) and found a way to fix this by mounting the ssl certificate, running "update-ca-certificates" and updating some environment variables for requests, node and httpx packages. Notice that you have to modify the container entry point for this.
Here is my docker-compose file:

``` yaml
version: "3.11"
services:
  litellm:
    build:
      context: .
      args:
        target: runtime
    image: ghcr.io/berriai/litellm:main-stable
    volumes:
      - ./config.yaml:/app/config.yaml
      - ./rootCA.pem:/usr/local/share/ca-certificates/rootCA.crt
    ports:
      - "4000:4000" # Map the container port to the host, change the host port if necessary
    environment:
        DATABASE_URL: "postgresql://POSTGRES_USER:POSTGRES_PASSWORD@litellm-postgres:5432/litellm"
        STORE_MODEL_IN_DB: "True" # allows adding models to proxy via UI
        LITELLM_MASTER_KEY: "LITELLM_MASTER_KEY"
        LITELLM_SALT_KEY: "LITELLM_SALT_KEY"
        NODE_EXTRA_CA_CERTS: "/etc/ssl/certs/ca-certificates.crt"
        SSL_CERT_PATH: "/etc/ssl/certs"
        REQUESTS_CA_BUNDLE: "/etc/ssl/certs/ca-certificates.crt"
        HTTPX_CA_BUNDLE: "/etc/ssl/certs/ca-certificates.crt"
        UI_USERNAME: "<ui user>"
        UI_PASSWORD: "<ui pw>"
    entrypoint: ["/bin/sh", "-c"] 
    command: ["update-ca-certificates && litellm --config /app/config.yaml --port 4000 --num_workers 4"]
    networks:
      - llms-network
    restart: always
    container_name: litellm-api
    depends_on:
      - db

 
  db:
    image: postgres
    restart: always
    environment:
      POSTGRES_DB: litellm
      POSTGRES_USER: POSTGRES_USER
      POSTGRES_PASSWORD: POSTGRES_PASSWORD
    healthcheck:
      test: ["CMD-SHELL", "pg_isready -d litellm -U litellm"]
      interval: 1s
      timeout: 5s
      retries: 10
    volumes:
      - ./postgres_db_data:/var/lib/postgresql/data
    ports:
      - "5432:5432"
    networks:
      - llms-network
    container_name: litellm-postgres
  
  prometheus:
    image: prom/prometheus
    volumes:
      - prometheus_data:/prometheus
      - ./prometheus.yml:/etc/prometheus/prometheus.yml
    ports:
      - "9090:9090"
    command:
      - '--config.file=/etc/prometheus/prometheus.yml'
      - '--storage.tsdb.path=/prometheus'
      - '--storage.tsdb.retention.time=15d'
    restart: always
    networks:
      - llms-network
    container_name: litellm-prometheus

volumes:
  prometheus_data:
    driver: local

networks:
  llms-network:
    name: llms-network
    driver: bridge


# ...rest of your docker-compose config if any

```

**Comment by SebLz** [comment]
I've been experiencing the same issue (https://docs.litellm.ai/docs/proxy/deploy#2-setting-ssl-certification didn't work in my case) and found a way to fix this by mounting the ssl certificate, running "update-ca-certificates" and updating some environment variables for requests, node and httpx packages. Notice that you have to modify the container entry point for this.
Here is my docker-compose file:

``` yaml
version: "3.11"
services:
  litellm:
    build:
      context: .
      args:
        target: runtime
    image: ghcr.io/berriai/litellm:main-stable
    volumes:
      - ./config.yaml:/app/config.yaml
      - ./rootCA.pem:/usr/local/share/ca-certificates/rootCA.crt
    ports:
      - "4000:4000" # Map the container port to the host, change the host port if necessary
    environment:
        DATABASE_URL: "postgresql://POSTGRES_USER:POSTGRES_PASSWORD@litellm-postgres:5432/litellm"
        STORE_MODEL_IN_DB: "True" # allows adding models to proxy via UI
        LITELLM_MASTER_KEY: "LITELLM_MASTER_KEY"
        LITELLM_SALT_KEY: "LITELLM_SALT_KEY"
        NODE_EXTRA_CA_CERTS: "/etc/ssl/certs/ca-certificates.crt"
        SSL_CERT_PATH: "/etc/ssl/certs"
        REQUESTS_CA_BUNDLE: "/etc/ssl/certs/ca-certificates.crt"
        HTTPX_CA_BUNDLE: "/etc/ssl/certs/ca-certificates.crt"
        UI_USERNAME: "<ui user>"
        UI_PASSWORD: "<ui pw>"
    entrypoint: ["/bin/sh", "-c"] 
    command: ["update-ca-certificates && litellm --config /app/config.yaml --port 4000 --num_workers 4"]
    networks:
      - llms-network
    restart: always
    container_name: litellm-api
    depends_on:
      - db

 
  db:
    image: postgres
    restart: always
    environment:
      POSTGRES_DB: litellm
      POSTGRES_USER: POSTGRES_USER
      POSTGRES_PASSWORD: POSTGRES_PASSWORD
    healthcheck:
      test: ["CMD-SHELL", "pg_isready -d litellm -U litellm"]
      interval: 1s
      timeout: 5s
      retries: 10
    volumes:
      - ./postgres_db_data:/var/lib/postgresql/data
    ports:
      - "5432:5432"
    networks:
      - llms-network
    container_name: litellm-postgres
  
  prometheus:
    image: prom/prometheus
    volumes:
      - prometheus_data:/prometheus
      - ./prometheus.yml:/etc/prometheus/prometheus.yml
    ports:
      - "9090:9090"
    command:
      - '--config.file=/etc/prometheus/prometheus.yml'
      - '--storage.tsdb.path=/prometheus'
      - '--storage.tsdb.retention.time=15d'
    restart: always
    networks:
      - llms-network
    container_name: litellm-prometheus

volumes:
  prometheus_data:
    driver: local

networks:
  llms-network:
    name: llms-network
    driver: bridge


# ...rest of your docker-compose config if any

```

**Pull request closed without merging**